[Nicholas Loo] ip

nwhloo99 opened the pull request:
# DukeBot
>“Your mind is for having ideas, not holding them.” – David Allen ([Source](https://dansilvestre.com/productivity-quotes/)) 

DukeBot frees your mind of having to remember things you need to do. It's,
* text-based
* easy to learn
* ~FAST~ _SUPERFAST_ to use

All you need to do is,
1. Download it from [here](https://github.com/nwhloo99/ip/releases/tag/v0.1).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉 

And it is **FREE!**

Features:

- [x] Managing Tasks
- [x] Managing Deadlines
- [x] Managing Events
- [x] Unit Testing
- [x] GUI
- [ ] Reminders (Coming soon)

If you are a Java programmer, you can use it to practice Java too. Here is the `Main` method:
```java
public class Launcher {
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
}
```

Commit:
Add Gradle support

nwhloo99 committed:
Added getCommand() method to Duke class to allow bot to read user inputs

nwhloo99 committed:
Added more ability to store commands in a list and to list the stored commands in getCommand()

nwhloo99 committed:
Created Task clas and modified getCommand() method to collect list of Task instead of list of Strings for commands

nwhloo99 committed:
Allowed user to mark tasks as done, and to view status when asking for list

nwhloo99 committed:
Moved blocks of code within the if else statements of getCommand() into their own functions to clean up code

nwhloo99 committed:
Created ToDO class, made Task class abstract

nwhloo99 committed:
Created Deadline Class

nwhloo99 committed:
Created Event class, slight modification to toString of Deadline

nwhloo99 committed:
Created functions to handle adding Todo, Deadline, Event in Duke class

nwhloo99 committed:
Formatted toString in Deadline and Event. Finished functions for adding Todo, Deadline and Event to list in Duke

nwhloo99 committed:
Created DukeException class

nwhloo99 committed:
Added exception handling for when invalid command was given

nwhloo99 committed:
Added exception handling for done command

nwhloo99 committed:
Added exception handling for todo command

nwhloo99 committed:
Added exception handling for adding deadline

nwhloo99 committed:
Added exception handling for adding event, added more check cases to adding deadline

nwhloo99 committed:
Added delete function

nwhloo99 committed:
Turned all the available commands into an Enum (DukeCommands). Created an interface DukeFunctions to handle the functions associated with the commands.

nwhloo99 committed:
Added automated testing, fixed some message formatting

nwhloo99 committed:
Add toFile() method to Task and its child classes

Task classes (e.g. Event, Deadline and Task) all need to be written to
duke.txt to be stored.

Giving each Task child class their own method follows OOP principles.

nwhloo99 committed:
Add DukeFileIO.java and SAVE Enum to DukeCommands

Methods are required to read/write data from/to duke.txt.

Creating DukeFileIO.java with methods to read current data from duke.txt
allows us to read past data from the user. Creating new Enum SAVE allows
users to instruct Duke to write all current data in list to duke.txt.

nwhloo99 committed:
Reading past data: removed extra space created

Reading past data should store the previous data in the list in the same
way.

Fixing the extra space created allows users to have a consistent view on
their list of tasks.

nwhloo99 committed:
Remove duke.txt for testing

nwhloo99 committed:
Add createDataFile() method to DukeFileIO class

In the case that duke.txt is missing from the user's data folder, a new
duke.txt file should be created for them.

Creating this createDataFile() method, ensures that the user will always
have the duke.txt file to use for data storage.

nwhloo99 committed:
Add createDataFile() method to DukeFileIO class

In the case that duke.txt is missing from the user's data folder, a new
duke.txt file should be created for them.

nwhloo99 committed:
Add createDataFile() method to DukeFileIO class

In the case that duke.txt is missing from the user's data folder, a new
duke.txt file should be created for them.

nwhloo99 committed:
Merge branch 'branch-level-7' of https://github.com/nwhloo99/ip into branch-level-7

nwhloo99 committed:
Add ability to store time in Deadline and Event as LocalDate

Deadline and Event are Tasks that track dates.

Adding the ability to store time as LocalDate, enforces that users will
have standardized storage of Event and Deadline in their lists.

nwhloo99 committed:
Merge branch 'branch-level-8'

# Conflicts:
#	src/main/java/DukeCommands.java

nwhloo99 committed:
Create TaskList class

nwhloo99 committed:
Create Storage class to handle saving/loading tasks to/from file

nwhloo99 committed:
Create Ui, Parser, Command classes to handle reading and outputting

nwhloo99 committed:
Create AddCommand class that inherits from Command

nwhloo99 committed:
Create DoneCommand class which inherits from Command

nwhloo99 committed:
Create ListCommand class which inherits from Command

nwhloo99 committed:
Update ListCommand

nwhloo99 committed:
Create ExitCommand class which inherits from Command

nwhloo99 committed:
Create SaveCommand class which inherits from Command

nwhloo99 committed:
Create DeleteCommand class which inherits from Command

nwhloo99 committed:
Update DeleteCommand class

nwhloo99 committed:
Create InvalidCommand class which inherits from Command

nwhloo99 committed:
Update parse method in Parser to detect commands

nwhloo99 committed:
Remove redundant constructors from Command subclasses

nwhloo99 committed:
Update ListCommand execute() method

nwhloo99 committed:
Update SaveCommand execute() method

nwhloo99 committed:
Update Parser parse() method

nwhloo99 committed:
Update Parser class

nwhloo99 committed:
Update TaskList toString() method

nwhloo99 committed:
Update name of TaskList variables to tasks

nwhloo99 committed:
Update DeleteCommmand and DoneCommand

nwhloo99 committed:
Updated Parser to not handle index out of bounds for delete and done

nwhloo99 committed:
Update TaskList toString() to display message if no elements

nwhloo99 committed:
Update Ui class to display better messages

nwhloo99 committed:
Clean up Dead Code/Redundant File

nwhloo99 committed:
Sort classes into packages

Number of classes have increased and organizing into packages tidies up
the project.

Sorting the classes into packages allows us to easily navigate between
different classes.

nwhloo99 committed:
Create Test methods in TaskListTest

Output from TaskList should be consistent and correct.

Testing TaskList methods outputs allows us to ensure that we always get
the result that we are expecting from TaskList.

nwhloo99 committed:
Update .gitIgnore

nwhloo99 committed:
Revert "Update .gitIgnore"

This reverts commit 4f020b42e6e75fd232db28fc830c439d1fa30d48.

nwhloo99 committed:
Revert "Revert "Update .gitIgnore""

This reverts commit 139a4c08f0212a1442ac7d393bb34f443dbb46c1.

nwhloo99 committed:
Add documentation for public methods in package duke.main

nwhloo99 committed:
Add Documentation for Command and its subclasses

nwhloo99 committed:
Add Documentation for Task and its subclasses

nwhloo99 committed:
Add getDescription method to Task

Description of Task is required to compare to other Strings to find
specific Task.

Adding the getDescription method to Task allows the TaskList to get
the Description of Task in its ArrayList to find which Task matches any
keyword given.

nwhloo99 committed:
Add findMatchingTask method to TaskList

Method is required for user to find specific Task matching the keyword
given.

Adding findMatchingTask method ensures that users can find the tasks in
the TaskList that match their keyword given.

nwhloo99 committed:
Create FindCommand class which inherits from Command

nwhloo99 committed:
Update Parser to handle find command

nwhloo99 committed:
Merge branch 'branch-A-JavaDoc'

nwhloo99 committed:
Merge branch 'add-gradle-support' of https://github.com/nwhloo99/ip into branch-A-Gradle

nwhloo99 committed:
Fix path to Main application

nwhloo99 committed:
Remove Checkstyle from build.gradle

Removing checkstyle allows the build function of gradle to run as the
program does not support checkstyle currently.

nwhloo99 committed:
Add checkstyle support

Adding checkstyle support will allow grade to automate checking of
styling of code to ensure that the code follows convention.

nwhloo99 committed:
Fix checkstyle errors

Fixing checkstyle errors ensures that the code follow coding convention.

nwhloo99 committed:
Update build.gradle to allow for JavaFx

Enabling JavaFx allows creating of a GUI to better the user experience
when using the application.

nwhloo99 committed:
Create base GUI that will echo user input

nwhloo99 committed:
Update the methods in Ui to return a String.

Returning a String allows these methods to pass a String to Label in the
GUI, which allows for the displaying of messages in the GUI.

nwhloo99 committed:
Add Getters to Duke

Adding getters to Duke allows us to access the private fields in Duke
which are required for the GUI.

nwhloo99 committed:
Add greet method to MainWindow

Adding greet method to MainWindow allows us to send the user a greeting.

nwhloo99 committed:
Send greeting message upon launching program

nwhloo99 committed:
Update execute method in Command classes

Returning Strings in execute methods allows us to pass a String to the
GUI components to display a message to the user.

nwhloo99 committed:
Clean up Duke

Removing methods that are no longer in use ensures that the code will be
cleaner.

nwhloo99 committed:
Update Ui methods

nwhloo99 committed:
Update DialogBox format

nwhloo99 committed:
Truncate Invalid command error message

Truncating the Invalid command error message ensures that the message
can fit nicely into the DialogBox

nwhloo99 committed:
Update handleUserInput() method in MainWindow

Updating the method allows the GUI to read user input and process it,
to find the correct command to output to the user.

nwhloo99 committed:
Remove redundant Test files

nwhloo99 committed:
Fix String format for displayAddedMessage in Ui

kevinmingtarja reviewed (commented):
LGTM! Good work and all the best with the remaining tasks :)

GabrielWLM commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,38 @@
+        this.task = task;
+    }
+
+    /**
+     * Executes the add command, which adds a Task to the list.
+     * @param tasks The list of Task.
+     * @param ui The Ui objects that handles input from user and output to user.
+     * @param storage The Storage object that handles reading/writing of data.
+     * @return String The message to display once add command has been completed.
+     * @throws DukeException Unused.
+     */
```
Perhaps you should add an empty line between the description and parameter section of the comment? I notice this in a few other places as well.

GabrielWLM commented on src/main/java/duke/gui/MainWindow.java:
```diff
@@ -0,0 +1,65 @@
+package duke.gui;
+
+import duke.command.Command;
+import duke.main.Duke;
+import duke.main.DukeException;
+import duke.main.Parser;
+import javafx.fxml.FXML;
+import javafx.scene.control.Button;
+import javafx.scene.control.Label;
+import javafx.scene.control.ScrollPane;
+import javafx.scene.control.TextField;
+import javafx.scene.layout.AnchorPane;
+import javafx.scene.layout.VBox;
```
Perhaps you could separate this group of import statements into smaller groups for readability? I notice the same thing in other places as well.

GabrielWLM commented on src/main/java/duke/main/Parser.java:
```diff
@@ -0,0 +1,186 @@
+     */
+    public static Command parse(String command) throws DukeException {
+        String first = command.toLowerCase().split(" ")[0];
+        switch (first) {
+        case "bye":
+            return new ExitCommand();
+        case "list":
+            return new ListCommand();
+        case "todo":
+            return parseToDo(command);
+        case "deadline":
+            return parseDeadline(command);
+        case "event":
+            return parseEvent(command);
+        case "delete":
+            return parseDelete(command);
+        case "done":
+            return parseDone(command);
+        case "save":
+            return new SaveCommand();
+        case "find":
+            return parseFind(command);
```
Is there a reason why you did not add a `//Fallthrough` comment for the cases where you do not include a break statement?

GabrielWLM reviewed (commented):
Nice Job! Your code looks readable and well organised.

Besides some formatting issues, I like that your code is very clean and organised. Great Work!

rebeccalaujx commented on src/main/java/duke/main/Storage.java:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    /**
+     * Method which searches for the data file and reads each line of the file, converting it into an ArrayList of Task
```
Perhaps start method headers with a verb like "Searches for..."

rebeccalaujx commented on src/main/java/duke/main/Storage.java:
```diff
@@ -0,0 +1,111 @@
+     * @return ArrayList The ArrayList of tasks that was stored in the data file.
+     * @throws FileNotFoundException In the case that the file does not exist.
+     */
+    public ArrayList<Task> readPastData() throws FileNotFoundException {
```
This method might be a little too long (>30 LOC), perhaps it might be better to include the cases in a separate method?

rebeccalaujx commented on src/main/java/duke/main/Storage.java:
```diff
@@ -0,0 +1,111 @@
+        return savedTasks;
+    }
+
+    private boolean checkIfDone(String component) {
```
Could perhaps be shortened to `return (Integer.parseInt(component.substring(1, 2)) == 1)`?

rebeccalaujx commented on src/main/java/duke/main/TaskList.java:
```diff
@@ -0,0 +1,114 @@
+     * @return TaskList The resultant TaskList that contains all the Task that match the keyword.
+     */
+    public TaskList findMatchingTask(String keyword) {
+        ArrayList<Task> resultList = new ArrayList<>();
```
Not sure if "list" is allowed but the coding standard mentions that plural form should be used on names representing a collection of objects, so perhaps "results" might be better? Same thing for testList below. Do correct me if I'm wrong though!

rebeccalaujx reviewed (commented):
LGTM! Keep up the good work 👍

nwhloo99 committed:
Change mainClassName in build.grade to duke.gui.Launcher

nwhloo99 committed:
Remove redundant if else statement in Storage

Redundant if else statement blocks may be confusing.

Removing the redundant if else statement improves the readability of the
code.

As a step towards improving code qualify, let's remove redundant if else
statement blocks to make subsequent modification of the code easier.

nwhloo99 committed:
Extract methods from case segments in readPastData() method in Storage

Previous code was very long and hard to read.

Extracting methods improves the readability of code and ensures that
future modifications to any of these methods would be easier.

As a step towards better code quality, let's extract those code blocks
into separate methods. Doing so will make subsequent modifications
easier and cleaner.

nwhloo99 committed:
Add assertions for displayAddedMessage() method in Ui

This method should never receive an empty Task or empty TaskList.

Adding assertions ensures that the method will always received valid
arguments.

Let's use assertions to ensure that the arguments are valid.

nwhloo99 committed:
Add assertion to displayCompletedMessage method in Ui

The method should never receive a null value for its Task argument.

Asserting that the Task received is not null will ensure that the
method will always work.

Let's use an assertion to ensure that the method will work.

nwhloo99 committed:
Added assertion to displayListOfTasks method in Ui

nwhloo99 committed:
Remove redundant arguments from displayDeleteMessage method in Ui

Redundant arguments may lead to confusion.

Removing the redundant arguments improves the readability of the code.

Let's remove redundant code segments as an effort to improve the code
quality of the application.

nwhloo99 committed:
Merge pull request #1 from nwhloo99/branch-A-Assertions

Branch-A-Assertions

nwhloo99 committed:
Merge branch 'master' into branch-A-CodeQuality

nwhloo99 committed:
Merge pull request #2 from nwhloo99/branch-A-CodeQuality

Branch-A-code-quality

nwhloo99 committed:
Bug Fix DeleteCommand execute method

nwhloo99 committed:
Create PriorityCommand to handle changing priority level

nwhloo99 committed:
Update Parser to handle priority commands

nwhloo99 committed:
Update Ui to return message for successful priority change

nwhloo99 committed:
Update Task and its subclasses to handle Priority

nwhloo99 committed:
Update format for writing to file

nwhloo99 committed:
Update read data file methods in Storage

Data now contains the priority level of Tasks.

Updating the read methods in Storage allows for the proper reading of
the past user data.

nwhloo99 committed:
Tweak DialogBox size and color

nwhloo99 committed:
Update GUI to show Red text from Duke when error has occured

nwhloo99 committed:
Add Title to GUI window

nwhloo99 committed:
Add Ui Screenshot

nwhloo99 committed:
Update format of add message

nwhloo99 committed:
Update format of add message

nwhloo99 committed:
Fix wrong message being displayed in PriorityCommand

nwhloo99 committed:
Update README to contain UG

nwhloo99 committed:
Update TaskListTest to include priority

nwhloo99 committed:
Add background to ScrollPane area in the GUI

nwhloo99 committed:
Add Ui picture example

nwhloo99 committed:
Fix Storage not reading priority properly

nwhloo99 committed:
Update filePath for jar File

nwhloo99 committed:
Add Method to allow Storage to create a folder if required

nwhloo99 committed:
Update README

nwhloo99 committed:
Update JavaDocs to be in imperative mood
